gnome-initial-setup/pages/site/gis-site-search-entry.c

```diff
@@ -0,0 +1,541 @@
+  g_object_unref (completion);
+
+  g_signal_connect (entry, "changed",
+                    G_CALLBACK (entry_changed), NULL);
+}
+
+static void
+finalize (GObject *object)
+{
+  GisSiteSearchEntryPrivate *priv = gis_site_search_entry_get_instance_private (
+    GIS_SITE_SEARCH_ENTRY (object));
+
+  g_clear_object (&priv->site);
+  g_clear_pointer (&priv->sites, g_ptr_array_unref);
+  g_clear_object (&priv->model);
+
+  G_OBJECT_CLASS (gis_site_search_entry_parent_class)->finalize (object);
+}
+
+static void
+gis_site_search_entry_class_init (
+        GisSiteSearchEntryClass *site_search_entry_class)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (site_search_entry_class);
+
+  object_class->finalize = finalize;
+  object_class->set_property = set_property;
+  object_class->get_property = get_property;
+
+  /* properties */
+  g_object_class_install_property (
+      object_class, PROP_SITES_AVAILABLE,
+      g_param_spec_boxed (GIS_SITE_SEARCH_ENTRY_PROP_SITES_AVAILABLE,
+                          "Full site list",
+                          "The array of sites available",
+                          G_TYPE_PTR_ARRAY,
+                          G_PARAM_WRITABLE));
+
+  g_object_class_install_property (
+      object_class, PROP_SITE,
+      g_param_spec_object (GIS_SITE_SEARCH_ENTRY_PROP_SITE,
+                           "Site",
+                           "The selected site",
+                           GIS_TYPE_SITE,
+                           G_PARAM_READWRITE));
+}
+
+static void
+set_property_sites_available (GisSiteSearchEntry *self,
+                              const GValue       *value)
+{
+  GisSiteSearchEntryPrivate *priv = gis_site_search_entry_get_instance_private (
+    self);
+  GtkListStore *store = NULL;
+  GtkEntryCompletion *completion;
+
+  store = gtk_list_store_new (3, G_TYPE_STRING, GIS_TYPE_SITE, G_TYPE_STRING);
+
+  priv->sites = g_value_dup_boxed (value);
+  for (int i = 0; priv->sites && i < priv->sites->len; i++)
+    {
+      GisSite *site = g_ptr_array_index (priv->sites, i);
+      fill_store (site, store);
+    }
+
+  g_clear_object (&priv->model);
+  priv->model = GTK_TREE_MODEL (store);
+
+  completion = gtk_entry_get_completion (GTK_ENTRY (self));
+  gtk_entry_completion_set_match_func (completion, matcher, NULL, NULL);
+  gtk_entry_completion_set_model (completion, GTK_TREE_MODEL (store));
+}
+
+static void
+set_property (GObject *object, guint prop_id,
+              const GValue *value, GParamSpec *pspec)
+{
+  GisSiteSearchEntryPrivate *priv = gis_site_search_entry_get_instance_private (
+    GIS_SITE_SEARCH_ENTRY (object));
+
+  switch (prop_id)
+    {
+      case PROP_SITES_AVAILABLE:
+        set_property_sites_available (GIS_SITE_SEARCH_ENTRY (object), value);
+        break;
+      case PROP_SITE:
+        gis_site_search_entry_set_site (GIS_SITE_SEARCH_ENTRY (object),
+                                        g_value_dup_object (value));
```

Comment:
You have to use `g_value_get_object()` here, since `gis_site_search_entry_set_site()` adds a new reference